### ISSUE TITLE
[FEATURE] streamz as an input

### ISSUE THREAD

**Issue by martindurant** [scikit-multiflow/scikit-multiflow#106]
I came across this work at anacondacon, and am impressed at the set of iterative learning methods implemented. However, I feel that the *input* data streams are not something that has so far been particularly important for you in development.

The pure-python real-time events processing library [streamz](https://streamz.readthedocs.io/en/latest/) may be just the link that can stand between the data sources (text files, TCP/HTTP server, kafka, times random processes ...), doing some data cleaning and processing, and then passing the events into the algorithms developed here. Indeed, the library comes with hooks for real-time automatic plotting, which might suit your project very nicely on the output side too.

This issue is here to investigate to what extent the two projects might collaborate for mutual benefit.

**Comment by jacobmontiel** [comment]
Thanks for the information @martindurant

As you mention, data ingestion is something that we are looking forward to improving. I talked with some people at AnacondaCON regarding this topic and they also pointed to `streamz`. This is definitely worth investigating.

Could you elaborate on
> Indeed, the library comes with hooks for real-time automatic plotting

as we are also interested in improving in terms of visualization. I also talked with the guys involved in `pyviz` and I believe we are into something good here.

@jmread @abifet

**Comment by martindurant** [comment]
As far as visualisation goes, I would firstly point you to the following page of examples
https://hvplot.pyviz.org/user_guide/Streaming.html

With the advent of ["Panel"](https://panel.pyviz.org/), a widget-interaction and feedback library, you can now embed such output in complete dashboard layouts, and serve these as stand-alone apps or in-notebook.

**Comment by martindurant** [comment]
Some more examples using holoviews directly: http://holoviews.org/user_guide/Streaming_Data.html (the viz stack goes, low-level to high, bokeh-holoviews-hvplot-panel)

**Comment by jacobmontiel** [comment]
Those are really interesting examples and thanks for the viz stack explanation, it can be confusing at first. I think `Panel` is worth considering if it allows to easily display related information and (optionally) configure what to display.

A minor "concern" is that holoviews uses web pages to display information and some users might find this as an overhead. Although this might be just my own perception.

Nonetheless, I think we can focus first on `streamz` and later on visualization, but it is good to know the potential.

**Comment by martindurant** [comment]
Of course, certainly. Streamz has been around for a little while and there are some examples in its own docs and otherwise, but development had been pretty slow recently. However, we've picked this up at Anaconda, in the context of pyviz and [Intake](https://intake.readthedocs.io) (which is all about exposing data sources easily and to users using declarative or dynamic catalogues). In short, now is a very good time to get involved! I will be rewriting the documentation over the next couple of weeks to make the package architecture and contents more obvious.

For multiflow's case, I would imagine wrapping the algorithm classes you have, such that events coming from streamz (single or batched) are passed to the algorithm in question upon `emit`, and the outputs are either the current metric/residuals (when training) or the prediction. Care will be needed to keep things asynchronous, since I assume processing within the model takes more CPU time than the more typical small lambda functions. In this way, we can complete the data flow from input (which can be files gaining data with time) and outputting algorithm performance and predictions, all in real-time.

**Comment by remiadon** [comment]
Hi,

For a research projects we made skmulitflow and streamz work together, in order to
* generate batches of random data with `skmultiflow.data.random_rbf_generator`
* build a [streamz.dataframe.DataFrame](https://streamz.readthedocs.io/en/latest/dataframes.html#streaming-dataframes) with it
* visualize the Dataframe with [hvplot](https://hvplot.pyviz.org)

As a user of both skmultiflow and streamz, I definitely think it's worth investing in this direction.

I have a few questions though. 
@martindurant 
>  I would imagine wrapping the algorithm classes you have, such that events coming from streamz (single or batched) are passed to the algorithm in question upon emit

For my experiments I used to separate data generation/ingestion and algorithm initilisation/calls
My guideline was to use streamz as a "handler", mainly using calls to `map`.
This usually gives something looking like the following
```python
import numpy as np 
from streamz import Stream
from skmultiflow.trees.arf_hoeffding_tree import ARFHoeffdingTree

def get_random(n=10): 
    return np.random.random((n, 3)), np.random.randint(0, 2, n) 

s = Stream()
arf = ARFHoeffdingTree()
s.map(yield_random).map(lambda data: arf.partial_fit(data[0], data[1]))
s.emit(10)  # send a batch of 10 entries, along with their targets, to a partial_fit
```

This approach has two advantages 
* Gives a strict separation between data engineering and datascience, i.e how you ingest/preprocess and how you trigger ML 
* No black magic in the backend, adaptive machine learning models are just used for their flexibility and ability to be partially fitted, not for complete management over dataflows
--> following this mindset streamz would not be a dependance of skmultiflow, just a library commonly used with it

On the other hand having this low-level approach certainly have disadvantages. Can anyone explain the pros and cons ? It's not very clear to me for the moment

**Comment by martindurant** [comment]
I mean something like the following
```
import numpy as np
from streamz import Stream
from skmultiflow.trees.arf_hoeffding_tree import ARFHoeffdingTree

fn = 'mydata'
open(fn, 'w')


@Stream.register_api()
class multiflow(Stream):

    def __init__(self, upstream, model, **kwargs):
        self.model = model
        super().__init__(upstream, **kwargs)

    def update(self, x, who=None):
        self.model.partial_fit(x[0], x[1])
        self._emit(self.model.get_model_measurements)


def output(n=10):
    data = np.random.random((n, 3)).tolist(), np.random.randint(0, 2, n).tolist()
    outstr = "[%s, %s]\n" % data
    with open(fn, 'a') as f:
        f.write(outstr)


def read(line):
    import json
    return json.loads(line)


s = Stream.from_textfile('mydata')
arf = ARFHoeffdingTree()

s2 = s.map(read).multiflow(model=arf)
out = s2.sink_to_list()

s.start()
```

(run this in an interactive session, command line or notebook)

Every time now that you call `output()`, the model is updated, and output arrives as a new item in the list `out`. Note that this output ought to be something meaningful like the current loss metric (I couldn't find the right thing), and then it could be put into a dataframe for plotting - say the improvement in the metric versus time.

**Comment by martindurant** [comment]
Note that the `mutiflow` class *could* have been accomplished using `.accumulate` (which takes a state and applies data to it to produce a new state), but I thought this was cleaner.

**Comment by remiadon** [comment]
OK I see, as this issue has been tagged as a `good first issue`, I am OK to work on this and submit a PR

**Comment by martindurant** [comment]
I'd certainly be happy to see it!

**Comment by jacobmontiel** [comment]
Sounds good to me too, thanks for looking into this @remiadon

**Comment by martindurant** [comment]
Is there any more action that can be taken here? I'd be happy to talk more about this, and come up with streamz/multiflow examples, with a little help from the people here.
(I glanced at the referenced PR, but I'm not sure of the connection)

**Comment by remiadon** [comment]
@martindurant sure, I'll add some examples in the `skmultiflow/demos` folder. 
But it leads me to another issue : I was thinking of creating notebooks instead of just updating the existing demos. I guess I'll flag this as an issue

Anyway my question is : Do you want streamz to reference skmultiflow, or the other way around ?

My first guess is that skmultiflow should give examples of usage with streamz, because streamz is now (via this PR) a dependency (the opposite is untrue).

**Comment by remiadon** [comment]
I share with you the first features I tried to add:

## 1. Data Generation
```python
Source.from_generator(multiflow_generator) # create a streamz.Stream from skmultiflow.generator 
```

## 2. Concept Drift
providing a droppin replacement for the skmultflow.data.ConceptDriftGenerator class,
make it a Stream, taking a stream as input and slowly pulling data from the new stream 
```python
stream_1.concept_drift(stream_2, position=200, alpha=45)
```

## 3.partial_fit / predictions
For a first version, a Stream that takes both entries and targets as input, as mentioned above by @martindurant 

----------
The second feature is a bit harder, I was thinking of saving it for another PR

**Comment by martindurant** [comment]
Would it be worthwhile having some live time to talk about this? It seems like there's a lot we can talk over. You can contact me using mdurant at anaconda dot com.

**Comment by jacobmontiel** [comment]
My two cents:

Looking into what is currently available in `skmultiflow` and what `streamz` brings into the table I see (at a high level):

<img width="1309" alt="Stream-1" src="https://user-images.githubusercontent.com/17923265/57658965-ec5aab80-75e0-11e9-8ca1-e61cc1d6f907.png">

`streamz` provides a step forward to extend data ingestion capabilities without reinventing the wheel (nice). 

In this sense, my idea is to have `streamz` as the backbone of the data ingestion, processing, etc. It would be desirable to keep the `skmultiflow.data.Stream` class (rename it to avoid confusion with `streamz.Stream`). 

```
Stream
________________
+ stream_metadata:metadata
+ streamz_object:streamz.Stream
________________
+ next_sample()
+ has _more_samples()
+ restart()
.... 
```

The rationale for this includes:
1) `skmultiflow.data.Stream` provides the interfaces upon which multiple elements of `skmultiflow` rely on.
2) Stream metadata is rather important/useful on the learning side 
3) Keep the impact on the end user at the minimum (avoid as possible a new learning curve)

This does not mean that I am not open to a different approach, I am just bringing up these concerns.

----
Regarding examples/demos, that is something that we can work out for the benefit of both projects. Althoug we are missing examples in some parts of the code, it is something worht investing on. +1 for using notebooks, although they are a pain to maintain (version-control-wise), they provide users with functional examples.

Regarding the PR (@remiadon). I don't understand the comment about `partial_fit`/`predict`. Those are functionalities reserved for learning methods per design. We should keep *data operations* and *learning operations* separated.

Let me know what do you think @martindurant @remiadon

**Comment by remiadon** [comment]
@jacobmontiel I pretty much agree with all of this.
I raised an issue for replacing demos with notebooks

In the PR I proposed `skmultiflow.data.Stream` has been replaced by `skmultiflow.data.BaseGenerator`, because this is a base definition for how you generators (Agrawal, SEA, etc...) work.

------------------- 
> Stream metadata is rather important/useful on the learning side

That much ? To me the ideal world would be to have skmultiflow provide models which are compatible with stream learning, but that could also be fitted "manually" on batches of data, just like it is done in sklearn. This level of abstraction would allow a user to provide it own model, just by implementing partial_fit/predict, either it comes from skmultiflow, sklearn, or any other piece of code following the partial_fit/predict definition

In this context the role of skmultiflow would be to provide tools and helpers to better interface those models with streamz for proper pipelining in a streaming environment.

Another advantage of keeping skmultiflow algorithms unaware of streamz, which I have encountered while developing the PR, is that it makes unit testing easier. This way one can test the algorithm itself, without any import/call to streamz, which greatly simplify things.

------------ 
> We should keep data operations and learning operations separated

I think you are right, the closest way of doing this is by asking the user to provide a streaming dataframe, containing both entries and targets.
Explanation : 
* we should not try to reconciliate entries and targets coming in at different paces. This should be done by the user, who is in charge of bringing data into the learning method
* skumultiflow already did that by bringing data (formerly know as Streams, beware of confusion here), with both entries and targets in the same structure


> Keep the impact on the end user at the minimum (avoid as possible a new learning curve)

This is very important I agree. I am open to discuss it. I already though of a few ways of doing this, each having its pros and cons

**Comment by jacobmontiel** [comment]
> > Stream metadata is rather important/useful on the learning side
> 
> That much ? To me the ideal world would be to have skmultiflow provide models which are compatible with stream learning, but that could also be fitted "manually" on batches of data, just like it is done in sklearn.

I agree with this. The idea of having `fit`, `partial_fit` receiving data as `np.ndarrays` instead of the `Stream` object goes in this direction. We follow the `sklearn` approach of passing data directly so the user has more flexibility.

However, there are some benefits of keeping the metadata in the `Stream` class. For example, `nominal_attributes` is used in Hoeffding Trees to determine the "attribute observer" type (numerical or categorical). Although it is the user's responsibility to set this parameter, having the information available in the `Stream` object reduces the cognitive load for the user and the chances of error:

```python
stream = SomeStreamSource()
ht = HoeffdingTree(HoeffdingTree(nominal_attributes=stream.cat_features_idx))
```

**Comment by jacobmontiel** [comment]
Considering the discussion so far, I rather not rush this feature into the next release (0.3.0). I propose to target it for 0.4.0 so we have more time to put the different parts in place (notebooks, documentation, etc.).

Does this make sense?

**Comment by remiadon** [comment]
@jacobmontiel OK for me !

**Comment by gauthamkrishna-g** [comment]
@jacobmontiel Hi, any updates here on this thread?

**Comment by jacobmontiel** [comment]
Hi, unfortunately, there is currently no active effort on this.

**Comment by martindurant** [comment]
Following reading of https://github.com/scipy-conference/scipy_proceedings/pull/566 , I thought I would just put a ping here to remind the existence of these thoughts.
Note that, since this discussion, passing metadata along with values of a stream is now possible, and setting the starting point of streaming-dataframes (i.e., checkpoint and rehydrate later) is currently in PR.

**Comment by ahmed-masud** [comment]
@jacobmontiel  I may be able to start contributing to this.

**Comment by jmrozanec** [comment]
@jacobmontiel @martindurant @remiadon we got a similar request: to support ingesting Kafka events in order to stream them to scikit-multiflow. We started working on the requirement and are willing to take this issue, to support Kafka and other stream sources such as streamz. May I get assigned to this issue? 

Part of our ongoing effort was pushed to [this fork](https://github.com/jmrozanec/scikit-multiflow). We are willing to consider all ideas drafted here and get a working version we all agree on.

**Comment by remiadon** [comment]
Hi everyone, happy to hear this is still ongoing. I won't have much time to get involved into the code, but I can review if you like,

Rémi

**Comment by martindurant** [comment]
Note that streamz already has kafka as a source, so you can do two in one.

**Comment by jmrozanec** [comment]
@remiadon @martindurant thanks! Let's wait for @jacobmontiel thoughts on this. We will be glad to deliver the feature.
